### PR TITLE
MAID-2957: Implement the proposed new design

### PR DESCRIPTION
**Pull request by octol**
Initial work on implementing the proposed new design. Very rough work so far, so be kind!

Usually I'd probably have squashed some of these commits, but I kept them granular for now.

Missing (that I'm aware of) are primarily the factory functions as well as serialisation stuff.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -71,8 +71,42 @@ use serde::{de::DeserializeOwned, Serialize};
...
+const SIGNATURE_BYTES: usize = sign::SIGNATUREBYTES;
+const SYMMETRIC_KEY_BYTES: usize = secretbox::KEYBYTES;
+
+quick_error! {
+    /// This error is returned if encryption or decryption fail.
+    /// The encryption failure is rare and mostly connected to serialisation failures.
+    /// Decryption can fail because of invalid keys, invalid data, or deserialisation failures.
+    #[derive(Debug)]
+    pub enum Error {
+        /// Occurs when serialisation or deserialisation fails.
```
Comment (povilasb) [inline src/lib.rs:88]: I know this is mostly just copy paste from the bottom, but still there's inconsistent language use:
```
Occurs when serialisation or deserialisation fails.

vs


This error is returned if encryption or decryption fail.
```
I'm talking about fails vs fail :)

File: src/lib.rs [hunk 2]
```diff
@@ -92,76 +126,23 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
...
-#[derive(Serialize, Deserialize)]
-struct CipherText {
-    nonce: [u8; box_::NONCEBYTES],
-    ciphertext: Vec<u8>,
-}
+impl PublicEncryptKey {
+    /// Construct public key from bytes. Useful when it was serialized before.
+    pub fn from_bytes(public_key: [u8; PUBLIC_ENCRYPT_KEY_BYTES]) -> Self {
+        Self {
+            encrypt: box_::PublicKey::from_slice(&public_key).unwrap(),
```
Comment (povilasb) [inline src/lib.rs]: We use `unwrap!` macro from https://crates.io/crates/unwrap where applicable - it gives us better output on panic.

File: src/lib.rs [hunk 3]
```diff
@@ -92,76 +126,23 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
...
-            .enumerate()
-            .for_each(|(i, v)| full_len_key[i] = v);
-        full_len_key
-    }
-
-    /// Returns a public signing key.
-    #[cfg(not(feature = "mock"))]
-    pub fn public_sign_key(&self) -> PublicSignKey {
-        self.sign.0
+    /// For anyone who wants to store public key.
```
Comment (povilasb) [inline src/lib.rs]: "to store **the** public key`?

File: src/lib.rs [hunk 4]
```diff
@@ -190,36 +171,42 @@ impl PublicKeys {
...
-    /// Verifies the detached `signature`.
-    ///
-    /// Returns `true` if the signature is valid the `data` is verified.
-    pub fn verify_detached(&self, signature: &Signature, data: &[u8]) -> bool {
-        sign::verify_detached(&signature.signature, data, &self.sign)
-    }
+/// Store secret key in `Arc` to discourage cloning it. Makes it more secure by not duplicating
+/// encryption keys in memory.
+/// Use `encrypt_key_pair()` to generate public and private keys pair.
+// NOTE, serde by default doesn't allow to serialize `Arc`. We have to enable this feature
```
Comment (povilasb) [inline src/lib.rs]: It was meant as internal comment, hence I would remove it :)

File: src/lib.rs [hunk 5]
```diff
@@ -190,36 +171,42 @@ impl PublicKeys {
...
     }
+}
 
-    /// Verifies the detached `signature`.
-    ///
-    /// Returns `true` if the signature is valid the `data` is verified.
-    pub fn verify_detached(&self, signature: &Signature, data: &[u8]) -> bool {
-        sign::verify_detached(&signature.signature, data, &self.sign)
-    }
+/// Store secret key in `Arc` to discourage cloning it. Makes it more secure by not duplicating
```
Comment (povilasb) [inline src/lib.rs]: I would reword it to smth like "Reference counted secret encryption key used to decrypt data previously encrypted with `PublicEncryptKey`"

File: src/lib.rs [hunk 6]
```diff
@@ -190,36 +171,42 @@ impl PublicKeys {
...
-    pub fn public_keys(&self) -> &PublicKeys {
-        &self.public
+    /// Computes a shared secret from our secret key and the recipient's public key.
+    pub fn shared_secret(&self, their_pk: &PublicEncryptKey) -> SharedSecretKey {
+        let precomputed = Arc::new(box_::precompute(&their_pk.encrypt, &self.inner.encrypt));
+        SharedSecretKey { precomputed }
+    }
+
+    /// For anyone who wants to store secret key.
+    pub fn as_bytes(self) -> [u8; SECRET_ENCRYPT_KEY_BYTES] {
```
Comment (povilasb) [inline src/lib.rs]: Reword to smth like "Get the inner secret key representation."?

File: src/lib.rs [hunk 7]
```diff
@@ -244,13 +237,86 @@ impl SecretKeys {
...
+    };
+    let sec_enc_key = SecretEncryptKey {
+        inner: Arc::new(SecretEncryptKeyInner {
+            encrypt: encrypt_sk,
+        }),
+    };
+    (pub_enc_key, sec_enc_key)
+}
+
+/// Public signing key
```
Comment (povilasb) [inline src/lib.rs]: Docs don't say more than the struct name itself. How about adding a sentence or two what's this signing key about? There's some info at https://download.libsodium.org/doc/public-key_cryptography/public-key_signatures.html

**Review by povilasb — CHANGES_REQUESTED** [review]

**Comment by octol** [comment]
Added a few more commits, hopefully it should build and test correctly now.

Still working on reformulating some of the docs. Also need to redo the mocked `from_slice` functions as they currently just set zeros. Would be good to have a little unit test stats as well. 

And not to forget, switch to the `unwrap!` macro.

File: src/lib.rs (hunk 2 above)
Comment (octol) [reply src/lib.rs]: Currently in lib.rs that crate seems to be hidden behind `#[cfg(any(test, feature = "mock"))]`. So is it ok to just remove that line?

Comment (povilasb) [reply src/lib.rs]: yep :+1:

File: src/mock_crypto.rs [hunk 8]
```diff
@@ -71,6 +71,13 @@ pub(crate) mod crypto_impl {
                      Serialize)]
             pub(crate) struct PublicKey(pub(crate) [u8; PUBLICKEYBYTES]);
 
+            impl PublicKey {
+                // TODO: just to make it build
+                pub(crate) fn from_slice(_bs: &[u8]) -> Option<Self> {
```
Comment (povilasb) [inline src/mock_crypto.rs]: hm, do we really need `from_slice()`, if we have `from_bytes()`? :)

Comment (octol) [reply src/mock_crypto.rs]: I think these `from_slice` functions are mocking `rust_sodium` though?

Comment (povilasb) [reply src/mock_crypto.rs]: Yeah, you're right. I missed the file name, sorry :)

File: src/lib.rs (hunk 1 above)
Comment (octol) [reply src/lib.rs:88]: Fixed in the last set of commits

**Review by povilasb — DISMISSED** [review]
Looks good to me

File: src/lib.rs [hunk 9]
```diff
@@ -71,8 +71,42 @@ use serde::{de::DeserializeOwned, Serialize};
...
+            from()
+        }
+        /// Occurs when we can't decrypt a message or verify the signature.
+        DecryptVerify(_e: ()) {
+            display("error decrypting/verifying message")
+            from()
+        }
+        /// Occurs in case of an error during initialisation.
+        InitError(e: i32) {
+            display("error while initialising random generator")
```
Comment (Fraser999) [inline src/lib.rs]: I think this should be "error while initialising safe_crypto", since it's the whole lib being initialised.  Same applies to the doc comment for the `init()` function just below here.

File: src/lib.rs [hunk 10]
```diff
@@ -71,8 +71,42 @@ use serde::{de::DeserializeOwned, Serialize};
...
+        }
+        /// Occurs when we fail to derive encryption key from password.
+        DeriveKey {
+            display("error deriving encryption key from password")
+        }
+    }
+}
 
 /// Initialise random number generator for the key generation functions.
 pub fn init() -> Result<(), ()> {
```
Comment (Fraser999) [inline src/lib.rs]: I think it'd be better to return an actual `Error` here rather than the unit type - so e.g.
```rust
pub fn init() -> Result<(), Error> {
    crypto_impl::init().map_err(|_| Error::InitError(-1))
}
```

File: src/lib.rs [hunk 11]
```diff
@@ -92,76 +126,23 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
     hashing_impl::sha3_256(data)
 }
 
-/// Represents a set of public keys, consisting of a public signature key and a public
-/// encryption key.
+/// Use `gen_encrypt_keypair()` to generate a public and private key pair.
```
Comment (Fraser999) [inline src/lib.rs]: I think this doc comment should first state what `PublicEncryptKey` is rather than how to create one.  Also, we're using the term "secret" rather than "private" when describing crypto keys now.

File: src/lib.rs [hunk 12]
```diff
@@ -92,76 +126,23 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
...
-    precomputed: Arc<box_::PrecomputedKey>,
-}
-
-#[derive(Serialize, Deserialize)]
-struct CipherText {
-    nonce: [u8; box_::NONCEBYTES],
-    ciphertext: Vec<u8>,
-}
+impl PublicEncryptKey {
+    /// Construct public key from bytes. Useful when it was serialized before.
```
Comment (Fraser999) [inline src/lib.rs]: We try to stick to UK spellings rather than US wherever possible, so "serialised" rather than "serialized".

File: src/lib.rs [hunk 13]
```diff
@@ -92,76 +126,23 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
...
-            .enumerate()
-            .for_each(|(i, v)| full_len_key[i] = v);
-        full_len_key
-    }
-
-    /// Returns a public signing key.
-    #[cfg(not(feature = "mock"))]
-    pub fn public_sign_key(&self) -> PublicSignKey {
-        self.sign.0
+    /// For anyone who wants to store the public key.
```
Comment (Fraser999) [inline src/lib.rs]: Rather than a specific use-case, it might be better to just state that this converts the `PublicEncryptKey` into the raw underlying bytes or something like that?

Also, [a Rust convention](https://rust-lang-nursery.github.io/rust-clippy/v0.0.207/index.html#wrong_self_convention) is that a method like this which consume `self` should be named `into_...`.  I just noticed that [clippy is actually failing on this PR](https://travis-ci.org/maidsafe/safe_crypto/jobs/415933111) because of these `as_bytes()` methods (but it didn't show as a failure on the CI run because of a clippy bug which is resolved in a later version of clippy).

File: src/lib.rs [hunk 14]
```diff
@@ -190,36 +171,40 @@ impl PublicKeys {
     pub fn encrypt_anonymous_bytes(&self, plaintext: &[u8]) -> Vec<u8> {
```
Comment (Fraser999) [inline src/lib.rs]: I think `anonymously_encrypt()` and `anonymously_encrypt_bytes()` are clearer function names, but maybe confirm with @ustulation before making this change if you agree?

**Review by Fraser999 — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 13 above)
Comment (octol) [reply src/lib.rs]: Ah thanks, ok I've updated the names to `into_...`

File: src/lib.rs [hunk 15]
```diff
@@ -190,36 +174,40 @@ impl PublicKeys {
...
 
-    /// Verifies the detached `signature`.
-    ///
-    /// Returns `true` if the signature is valid the `data` is verified.
-    pub fn verify_detached(&self, signature: &Signature, data: &[u8]) -> bool {
-        sign::verify_detached(&signature.signature, data, &self.sign)
-    }
+/// Reference counted secret encryption key used to decrypt data previously encrypted with
+/// `PublicEncryptKey`.
+/// Use `gen_encrypt_keypair()` to generate a public and private key pair.
```
Comment (Fraser999) [inline src/lib.rs]: "private" → "secret"

File: src/lib.rs [hunk 16]
```diff
@@ -190,36 +174,40 @@ impl PublicKeys {
...
-            sign: sign_pk,
-            encrypt: encrypt_pk,
-        };
-        SecretKeys {
-            public,
-            inner: Arc::new(SecretKeysInner {
-                sign: sign_sk,
-                encrypt: encrypt_sk,
+impl SecretEncryptKey {
+    /// Construct secret key from given bytes. Useful when secret key was serialized before.
```
Comment (Fraser999) [inline src/lib.rs]: "serialized" → "serialised"

File: src/lib.rs [hunk 17]
```diff
@@ -244,13 +238,88 @@ impl SecretKeys {
...
+
+    /// Construct from bytes. Useful when it was serialised before.
+    pub fn from_bytes(public_key: [u8; PUBLIC_SIGN_KEY_BYTES]) -> Self {
+        Self {
+            sign: sign::PublicKey(public_key),
+        }
+    }
+
+    /// Convert the `PublicSignKey` into the raw underlying bytes.
+    /// For anyone who wants to store the public signature key.
```
Comment (Fraser999) [inline src/lib.rs]: "signature" → "signing"

File: src/lib.rs [hunk 18]
```diff
@@ -260,21 +329,41 @@ impl SecretKeys {
...
         }
     }
+}
 
-    /// Computes a shared secret from our secret key and the recipient's public key.
-    pub fn shared_secret(&self, their_pk: &PublicKeys) -> SharedSecretKey {
-        let precomputed = Arc::new(box_::precompute(&their_pk.encrypt, &self.inner.encrypt));
-        SharedSecretKey { precomputed }
-    }
+/// Construct random public/private signing key pair.
```
Comment (Fraser999) [inline src/lib.rs]: "public/private" → "public and secret"

File: src/lib.rs [hunk 19]
```diff
@@ -260,21 +329,41 @@ impl SecretKeys {
...
+        inner: Arc::new(SecretSignKeyInner { sign: sign_sk }),
+    };
+    (pub_sign_key, sec_sign_key)
 }
 
-impl Default for SecretKeys {
-    fn default() -> Self {
-        Self::new()
-    }
+/// Precomputed shared secret key.
```
Comment (Fraser999) [inline src/lib.rs:344]: I would break this doc comment block into two paragraphs (i.e. add a line with just `///` immediately after the first comment line) in order to keep the summary text fairly brief.

File: src/lib.rs [hunk 20]
```diff
@@ -343,13 +432,32 @@ impl SharedSecretKey {
...
 }
 
+/// Detached signature.
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, Clone)]
+pub struct Signature {
+    signature: sign::Signature,
+}
+
 impl Signature {
+    /// Construct the signature from bytes. Useful when it was serialized before.
```
Comment (Fraser999) [inline src/lib.rs]: "serialized" → "serialised"

File: src/lib.rs [hunk 21]
```diff
@@ -426,41 +547,42 @@ impl Default for SymmetricKey {
...
-            write!(f, "{:02x}", b)?;
-        }
-        for b in &self.encrypt[..] {
-            write!(f, "{:02x}", b)?;
-        }
+        write!(
+            f,
+            "{:02x}{:02x}{:02x}..",
+            &self.encrypt.0[0], &self.encrypt.0[1], &self.encrypt.0[2]
+        )?;
```
Comment (Fraser999) [inline src/lib.rs]: If you remove `?;` here, you don't need to return `Ok(())` below.  Same for other two `impl fmt::Display`s below.

File: src/lib.rs [hunk 22]
```diff
@@ -260,21 +329,41 @@ impl SecretKeys {
...
+/// Can be created from a pair of our secret key and the recipient's public key.
+/// As a result, we'll get the same key as the recipient with their secret key and
+/// our public key.
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone)]
+pub struct SharedSecretKey {
+    precomputed: Arc<box_::PrecomputedKey>,
 }
 
 impl SharedSecretKey {
+    /// Construct shared secret key from bytes. Useful when it was serialized before.
```
Comment (Fraser999) [inline src/lib.rs]: "serialized" → "serialised"

**Review by Fraser999 — DISMISSED** [review]

**Review by Fraser999 — APPROVED** [review]
